beluga/test/beluga/algorithm/test_cluster_based_estimation.cpp

```diff
@@ -333,8 +333,7 @@ TEST_F(ClusterBasedEstimationDetailTesting, ClusterEstimation) {
 
   const auto particles = ranges::views::zip(states, weights, clusters);
 
-  auto cluster_0_particles =
-      particles | ranges::views::cache1 | ranges::views::filter([](const auto& p) { return std::get<2>(p) == 0; });
+  auto cluster_0_particles = particles | ranges::views::filter([](const auto& p) { return std::get<2>(p) == 0; });
```

Comment:
The filter is there, I just removed the `cache1` adaptor, since it makes the resulting pipeline an input range regardless of the view category it is adapting.

Reply:
Yes, sorry, I meant the cache. If I remember correctly we needed it because otherwise the filter would evaluate the previous stage more than once, which may not always matter a lot if whatever is before is idempotent, but when it did matter, the effects tended to be rather surprising unless you knew what you look for.

I believe that that particular cache is there because I already spent time debugging this issue in some iteration of this code; removing it now does not break the test, but risks having someone else waste the same time some time from now.

So my question is, is the change worth it by some other measure?

Reply:
I hear the pain of using the filter view. I believe you are referring to [TPOIASI](https://www.fluentcpp.com/2019/02/12/the-terrible-problem-of-incrementing-a-smart-iterator/) (yay for more C++ acronyms :sweat_smile:).

I would have to measure again but *I think* that the compiler can see through this problem when everything is inlined if there are no side effects.

Also yes, `cache1` mitigates this issue when you are adding a print statement. But it (erroneously?) downgrades the input view into an input range when it could be more than that (forward, bidirectional or random access). I don't see a technical reason why it does that. This conflicts with the static assertions of our algorithms.

I'd rather have more correct code (and less template instantiations) when it is not strictly needed.

If we are interested, I can also implement [`cache_last`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2760r1.html#cache_last) as it is being considered for standardization provided we don't care as much about [`[res.on.data.races]`](https://wg21.link/res.on.data.races) (which we probably don't).

Reply:
Same question about materializing the view. If we intend our algorithms to perform no heap memory allocations, we should document that it comes at a cost.

Reply:
> I'd rather have more correct code (and less template instantiations) when it is not strictly needed.

SGTM.

Reply:
> Same question about materializing the view. If we intend our algorithms to perform no heap memory allocations, we should document that it comes at a cost.

This is unit test code. So I'm not sure how this applies here. But I will reply in the other thread about materialization.